Fix LDAP groups

lhuedepohl opened the pull request:
It seems the LDAP group feature is not heavily used, for me at least it always produced exceptions and I had to tinker with `api/app/models/user_ldap_strategy.rb` a bit to get it working. Note that I have absolutely no Ruby experience, so feel free to take this PR as a mere suggestion on how to potentially fix that feature.


Commit db20774368 fixes what seems to be actual errors with Ruby's scoping rules, but I am of course not sure. Without those, I at least always got errors of this kind:
```
[2022-05-18T11:56:15.549185 #17847] FATAL -- : [690f9afb-ffeb-470b-80ed-4e70254f4820] [17847:2752.05] ActionView::Template::Error (undefined local variable or method `login' for #<UserLdapStrategy:0x0000557bc522e9d0>
```
due to the missing `@` sigil for the (instance?) variable `login`, apparently.

Also,
```
E, [2022-05-10T13:13:57.737533 #3434] ERROR -- : [3434:29.02] Exception: undefined method `render_grouplist_ldap' for #<UserLdapStrategy:0x0000558e51f26450>
```
for the missing `UserLdapStrategy.`. Let me know if this is in any way the proper way to fix those.


The other commit, 865ae51e6d, fixes an issue when interacting with OpenLDAP, where some "operational" attributes are apparently not returned for an unqualified query for all attributes. You have to explicitly ask for them. The fix is to always ask for the configured `CONFIG['ldap_user_memberof_attr']` attribute, I believe this should be harmless and possible for all LDAP backends (but I have not tested this with an AD setup, for example).

Let me know if these suggestions help you, kind regards,
  Lorenz

vpereira commented on src/api/app/models/user_ldap_strategy.rb:
```diff
@@ -361,7 +361,7 @@ def relationship_groups_contains_user?(relationships, method_name)
       return false if relationship.group.nil?
       # check whether current user is in this group
-      return true if user_in_group_ldap?(login, relationship.group)
+      return true if user_in_group_ldap?(@login, relationship.group)
```
I don't think the instance variable is the problem here. I don't see it being initialized anywhere. `login` for sure wouldn't work either. I think what you want here is `User.session`

lhuedepohl replied:
As said, I have zero Ruby knowledge. With those changes, though, LDAP groups work for me - provided that I create them in OBS first, with the proper name. Not sure if this is intended behavior. Also, no members are listed in the "/groups" Web-UI page, but the membership checks seem to work correctly.

The feature certainly does not seem widely used, if at all ;)

lhuedepohl commented:
Just had to patch this in again when updating to 2.10.18. Based on your suggestion I changed

`return true if user_in_group_ldap?(@login, relationship.group)`

to

`return true if user_in_group_ldap?(User.session, relationship.group)`

which equally seems to work. Again - no Ruby experience here, but group permissions work with those two commits for us.

lhuedepohl commented:
@vpereira: There has now been another release meanwhile, is there anything that blocks a merge of this rather simple change?

Patching it manually after every update gets old at some point ;)

lhuedepohl commented:
I again had to apply this to our installation after the last release. Rebased to current master again (let me know if this is appropriate or you would rather that I not do this).


Also, please, let me know how I can help to get this merged. We use it like this in production, with a ~medium sized installation - the web UI says: _MPCDF Build Service hosts 71 projects, with 4,538 packages, in 3,721 repositories and is used by 32 confirmed developers._

eduardoj commented:
Hi @lhuedepohl!

First of all, sorry your PR got overlooked. Your PR fixes one issue we found and commented here: https://github.com/openSUSE/open-build-service/pull/16219#pullrequestreview-2098992472

After #16219 was merged, the branch of your PR has conflicts with master and needs a rebase. I'll be glad to take a look again to your PR after the conflicts are solved.

lhuedepohl committed:
Fix LDAP group handling

These two little changes made it possible for me to be able to use also
groups from LDAP, again

lhuedepohl committed:
LDAP groups: explicitly ask for member attribute

"Operational" attributes such as 'memberOf' are not returned by (at
least) openldap in a normal LDAP query, unless explictly requested.

lhuedepohl commented:
> Hi @lhuedepohl!
> 
> First of all, sorry your PR got overlooked. Your PR fixes one issue we found and commented here: [#16219 (review)](https://github.com/openSUSE/open-build-service/pull/16219#pullrequestreview-2098992472)
> 
> After #16219 was merged, the branch of your PR has conflicts with master and needs a rebase. I'll be glad to take a look again to your PR after the conflicts are solved.

Hi there!

I just needed to do this again after 2.10.25 was released. I also just read that LDAP will be discontinued (for a while) in the development branches, in order to get up-do-date LDAP modules set-up?

Thus, I don't think it makes sense to work on rebasing these changes, the file also looks completely different.

Nevertheless, I have a working file for 2.10.25, which I would be happy to share, if there is interest. At least it would work then meanwhile in the stable branch.

Ciao,
  Lorenz

lhuedepohl commented:
(Changed base to 2.10, in case there is interest)
